compiler/dcalc/from_scopelang.ml

```diff
@@ -307,9 +307,121 @@ let rec translate_expr (ctx : 'm ctx) (e : 'm Scopelang.Ast.expr) :
+                 field sc_sig.scope_sig_output_struct (Expr.with_ty m typ)
+             in
+             match Marked.unmark typ with
+             | TArrow (t_in, t_out) ->
+               (* Here the output scope struct field is a function so we
+                  eta-expand it and insert logging instructions. Invariant:
+                  works because user-defined functions in scope have only one
+                  argument. *)
+               let param_var = Var.make "param" in
+               let f_markings =
+                 [ScopeName.get_info scope; StructField.get_info field]
+               in
+               Expr.make_abs
+                 (Array.of_list [param_var])
+                 (tag_with_log_entry
+                    (tag_with_log_entry
+                       (Expr.eapp
+                          (tag_with_log_entry original_field_expr BeginCall
+                             f_markings)
+                          [
+                            tag_with_log_entry
+                              (Expr.make_var param_var (Expr.with_ty m t_in))
+                              (VarDef (Marked.unmark t_in))
+                              (f_markings @ [Marked.mark (Expr.pos e) "input"]);
+                          ]
+                          (Expr.with_ty m t_out))
+                       (VarDef (Marked.unmark t_out))
+                       (f_markings @ [Marked.mark (Expr.pos e) "output"]))
+                    EndCall f_markings)
+                 [t_in] (Expr.pos e)
+             | _ -> original_field_expr)
+           (StructName.Map.find sc_sig.scope_sig_output_struct ctx.structs))
+        (Expr.with_ty m (TStruct sc_sig.scope_sig_output_struct, Expr.pos e))
+    in
+    let if_then_else_returned =
+      Expr.eifthenelse
+        (tag_with_log_entry
+           (Expr.box
+              (Marked.mark
+                 (Expr.with_ty m (TLit TBool, Expr.pos e))
+                 (ELit (LBool true))))
+           PosRecordIfTrueBool direct_output_info)
+        (Expr.make_var result_eta_expanded_var
+           (Expr.with_ty m (TStruct sc_sig.scope_sig_output_struct, Expr.pos e)))
+        (Expr.box
+           (Marked.mark
+              (Expr.with_ty m
+                 (TStruct sc_sig.scope_sig_output_struct, Expr.pos e))
+              (ELit LEmptyError)))
+        (Expr.with_ty m (TStruct sc_sig.scope_sig_output_struct, Expr.pos e))
+    in
+    Expr.make_let_in result_var
+      (TStruct sc_sig.scope_sig_output_struct, Expr.pos e)
+      calling_expr
+      (Expr.make_let_in result_eta_expanded_var
+         (TStruct sc_sig.scope_sig_output_struct, Expr.pos e)
+         result_eta_expanded
+         (tag_with_log_entry
+            (tag_with_log_entry if_then_else_returned
+               (VarDef (TStruct sc_sig.scope_sig_output_struct))
+               direct_output_info)
+            EndCall
+            [ScopeName.get_info scope; Marked.mark (Expr.pos e) "direct"])
+         (Expr.pos e))
+      (Expr.pos e)
```

Comment:
Good call on the workaround! My bad on completely overlooking the logging issue when I added the scope calls. The solution doesn't feel very satisfying, but it should be solid enough, and hopefully we can have some simplification when we move the function call logs to the function themselves, as you are suggesting.

Reply:
Yes